[Core] Unify function schema parsing

hinthornw opened the pull request:
Use pydantic to infer nested schemas and all that fun.
Include bagatur's convenient docstring parser
Include annotation support


Previously we didn't adequately support many typehints in the bind_tools() method on raw functions (like optionals/unions, nested types, etc.)

hinthornw committed:
[Core] Unify function parsing

hinthornw committed:
Lint pt 1

hinthornw committed:
Fmt pt 1

hinthornw committed:
Lint pt 2

hinthornw committed:
Lint pt 4

eyurtsev commented on libs/core/langchain_core/tools.py:
```diff
@@ -89,19 +123,83 @@ def _create_subset_model(
+    }
+
+
+def _parse_python_function_docstring(function: Callable) -> Tuple[str, dict]:
```
My opinion for the parser is that the default would've ideally been opt-in from the get-go AND it would've raised exceptions on incomplete doc-string.

Given we can't do that right now are we sure that we want opt-in by default? Instead of adding an option to fail loudly?

wwkk7080 committed:
update wrong words (#23749)

Thank you for contributing to LangChain!

- [ ] **PR title**: "package: description"
- Where "package" is whichever of langchain, community, core,
experimental, etc. is being modified. Use "docs: ..." for purely docs
changes, "templates: ..." for template changes, "infra: ..." for CI
changes.
  - Example: "community: add foobar LLM"


- [ ] **PR message**: ***Delete this entire checklist*** and replace
with
    - **Description:** a description of the change
    - **Issue:** the issue # it fixes, if applicable
    - **Dependencies:** any dependencies required for this change
- **Twitter handle:** if your PR gets announced, and you'd like a
mention, we'll gladly shout you out!


- [ ] **Add tests and docs**: If you're adding a new integration, please
include
1. a test for the integration, preferably unit tests that do not rely on
network access,
2. an example notebook showing its use. It lives in
`docs/docs/integrations` directory.


- [ ] **Lint and test**: Run `make format`, `make lint` and `make test`
from the root of the package(s) you've modified. See contribution
guidelines for more: https://python.langchain.com/docs/contributing/

Additional guidelines:
- Make sure optional dependencies are imported within a function.
- Please do not add dependencies to pyproject.toml files (even optional
ones) unless they are required for unit tests.
- Most PRs should not touch more than one package.
- Changes should be backwards compatible.
- If you are adding something to community, do not re-import it in
langchain.

If no one reviews your PR within a few days, please @-mention one of
baskaryan, efriis, eyurtsev, ccurme, vbarda, hwchase17.

ccurme reviewed (commented):
A few thoughts:

IMO it could be nice to de-couple some of the changes here. Expanding the scope of docstring parsing worries me a little. Consider the scenario where docstrings are outdated, and tool calling was previously working in spite of the outdated docstrings (e.g., relying on type hints, which may be less likely to get outdated due to linting)-- could this change adversely impact users in that situation?

---

Using the tool parsing for functions feels like a straightforward win. There are edge cases (probably undocumented) where it could introduce errors. For example, this used to work, but now raises ValueError:
```python
from langchain_core.utils.function_calling import convert_to_openai_function

class float:
    pass

def my_function(x: float) -> None:
    pass

convert_to_openai_function(my_function)
```
I can't think of any errors that I'd consider very realistic scenarios though.

hinthornw commented:
I'm down to not parse docstrings by default / separate in another pr

hinthornw committed:
Rm docstring changes

hinthornw commented:
Git rebase 🤦
